File: backend/app/core/upgrades/stat_modifiers/damage_4.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Damage4(UpgradeImplementation):
    id = 'damage_4'
    name = 'Devastation'
    description = '+3.5% damage.'
    rarity = 'epic'
    type = 'stat_modifier'
    cost = 20

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.035
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/damage_4.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Damage4Def: UpgradeDef = {
  id: "damage_4",
  name: "Devastation",
  description: "+3.5% damage.",
  rarity: RarityID.Epic,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 20,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 0.035,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
}

export class Damage4 extends Upgrade {}
